### PR TITLE
access log add detour tag

### PR DESCRIPTION
**Pull request by fcying**
https://github.com/v2ray/v2ray-core/pull/1725

### CODE REVIEW

File: app/dispatcher/default.go [hunk 1]
```diff
@@ -265,11 +267,21 @@ func (d *DefaultDispatcher) routedDispatch(ctx context.Context, link *transport.
...
 			}
+			detourTag = tag
 		} else {
 			newError("default route for ", destination).WriteToLog(session.ExportIDToError(ctx))
 		}
 	}
 
+	accessMessageCount := len(log.AccessMessageChan)
+	if (accessMessageCount != 0) {
+		accessMessage := <-log.AccessMessageChan
```
Comment (eycorsican) [inline app/dispatcher/default.go]: V2Ray dispatches requests concurrently, you may get the wrong access message under high workload. A better way to implement this feature is to associate the access message to `ctx`.

Comment (fcying) [reply app/dispatcher/default.go]: I rewrite it by using `ctx`, please check it again.

File: common/session/context.go [hunk 2]
```diff
@@ -56,3 +58,14 @@ func ContentFromContext(ctx context.Context) *Content {
 	}
 	return nil
 }
+
+func ContextWithAccessMessage(ctx context.Context, accessMessage *log.AccessMessage) context.Context {
+	return context.WithValue(ctx, accessMessageKey, accessMessage)
+}
+
+func ContentFromAccessMessage(ctx context.Context) *log.AccessMessage {
```
Comment (eycorsican) [inline common/session/context.go]: Better rename to AccessMessageFromContext() according to the naming convention, and put them to the common/log package.

File: app/dispatcher/default.go [hunk 3]
```diff
@@ -265,11 +267,18 @@ func (d *DefaultDispatcher) routedDispatch(ctx context.Context, link *transport.
...
+			detourTag = tag
 		} else {
 			newError("default route for ", destination).WriteToLog(session.ExportIDToError(ctx))
 		}
 	}
 
+	accessMessage := session.ContentFromAccessMessage(ctx)
+	if accessMessage != nil {
+		accessMessage.Detour = "[" + detourTag + "]"
+		log.Record(accessMessage)
```
Comment (eycorsican) [inline app/dispatcher/default.go]: Should do the logging after all checks, and get the default outbound tag by handler.Tag().

File: proxy/socks/server.go [hunk 4]
```diff
@@ -117,7 +117,7 @@ func (s *Server) processTCP(ctx context.Context, conn internet.Connection, dispa
 		dest := request.Destination()
 		newError("TCP Connect request to ", dest).WriteToLog(session.ExportIDToError(ctx))
 		if inbound != nil && inbound.Source.IsValid() {
-			log.Record(&log.AccessMessage{
+			ctx = session.ContextWithAccessMessage(ctx, &log.AccessMessage{
```
Comment (eycorsican) [inline proxy/socks/server.go]: Do the same thing to the udp part

File: proxy/shadowsocks/server.go [hunk 5]
```diff
@@ -175,7 +175,7 @@ func (s *Server) handleConnection(ctx context.Context, conn internet.Connection,
 	inbound.User = s.user
 
 	dest := request.Destination()
-	log.Record(&log.AccessMessage{
+	ctx = session.ContextWithAccessMessage(ctx, &log.AccessMessage{
```
Comment (eycorsican) [inline proxy/shadowsocks/server.go]: Do the same thing to the udp part

File: proxy/socks/server.go (hunk 4 above)
Comment (fcying) [reply proxy/socks/server.go]:
```
common/mux/server.go:114:       msg := &log.AccessMessage{
```
mux/server.go have one also, Does this need to be modified?

File: common/session/context.go (hunk 2 above)
Comment (fcying) [reply common/session/context.go]: put `ContextWithAccessMessage` and `AccessMessageFromContext` to common/log package, use a public `AccessMessageKey`? if move `AccessMessageKey` to common/log, May be inconvenient to manage context key(I am not sure if the key can use the same value.)?

Comment (eycorsican) [reply common/session/context.go]: You create a new set of keys, see common/protocol/context.go

File: proxy/socks/server.go (hunk 4 above)
Comment (eycorsican) [reply proxy/socks/server.go]: Not quite sure at first glance, you could test if it works after making the changes.

Comment (fcying) [reply proxy/socks/server.go]: I don't know what mux is for. I haven't set this up, this is still not changed. Others have been changed.

**Comment by eycorsican** [comment]
LGTM

**Comment by kslr** [comment]
I will merge this commit after ray sync.

**Pull request merged**